### PR TITLE
WIP aarch64 sandbox (configurable sandbox+toolchain)

### PR DESCRIPTION
**Pull request by mahaase**
hey guys, i need help! :)

goal is to bring up a configurable sandbox+toolchain.
i want to setup an aarch64 sandbox incl. toolchain.
many stuff is working out of the box, including the aarch64 sandbox, if qemu-aarch64-static is installed.

NOTE: u should use qemu5. older versions have very bad performance! normally u should build it urself. default version of ubuntu18 lts is 2.11... sooooo far behind.

```
./dev/build/libs/compat/glibc-dev/1/build.sh shell
bash-4.3$ uname -a
Linux bob 4.15.0-135-generic #139-Ubuntu SMP Mon Jan 18 17:38:24 UTC 2021 aarch64 GNU/Linux
```

there was nothing special i had to do.

general: x86_64 host system -> cross-compiling am aarch64 sandbox -> than we are on a host-aarch64 system, with which we have to go forward to build the project.

i have done already some fixes to get some module buildable for cross-compiling.

this one is blocking the `TODO` marked stuff: https://github.com/BobBuildTool/bob/issues/399

what do u think about this?

br.

### CODE REVIEW

File: recipes/libs/expat.yaml [hunk 1]
```diff
@@ -1,14 +1,15 @@
 inherit: [autotools]
 
 metaEnvironment:
-   PKG_VERSION: "2.1.0"
+   PKG_VERSION: "2.2.10"
```
Comment (mahaase) [inline recipes/libs/expat.yaml:4]: includes the support for aarch64 targets.

File: recipes/libs/compat/glibc.yaml [hunk 2]
```diff
@@ -36,6 +36,7 @@ buildScript: |
             --with-headers=$2/usr/include \
             --enable-obsolete-rpc \
             --libdir=/usr/lib \
+            --disable-werror \
```
Comment (Ferruck) [inline recipes/libs/compat/glibc.yaml:39]: Why is this needed? I'd rather add a patch than disabling werror altogether...

**Review by Ferruck — COMMENTED** [review]
As always, I'm more than confused when it comes to this `host-toolchain` vs. `target-toolchain` stuff. From my perspective it looks good, but I'd like to wait for @jkloetzke's opinion.

File: recipes/libs/compat/glibc.yaml (hunk 2 above)
Comment (mahaase) [reply recipes/libs/compat/glibc.yaml:39]: while cross-compiling for aarch64 i had an error: `-Werror=missing-attributes`
with `--disable-werror` i was able to build it successfully

**Comment by mahaase** [comment]
I get stucked, and no ideas so far how to fix this.

```
a - /bob/86066a2b539d685119e16ea8782838723e2bc5d1/workspace/build/time/rtld-setitimer.os
mv -f /bob/86066a2b539d685119e16ea8782838723e2bc5d1/workspace/build/elf/rtld-libc.aT /bob/86066a2b539d685119e16ea8782838723e2bc5d1/workspace/build/elf/rtld-libc.a
make[3]: Leaving directory '/bob/62d2054d522222a7f0518640dddd7e374c288ef8/workspace/elf'
aarch64-bob-linux-gnu-gcc   -nostdlib -nostartfiles -r -o /bob/86066a2b539d685119e16ea8782838723e2bc5d1/workspace/build/elf/librtld.os '-Wl,-(' /bob/86066a2b539d685119e16ea8782838723e2bc5d1/workspace/build/elf/dl-allobjs.os /bob/86066a2b539d685119e16ea8782838723e2bc5d1/workspace/build/elf/rtld-libc.a -lgcc '-Wl,-)' \
      -Wl,-Map,/bob/86066a2b539d685119e16ea8782838723e2bc5d1/workspace/build/elf/librtld.os.map
aarch64-bob-linux-gnu-gcc   -nostdlib -nostartfiles -shared -o /bob/86066a2b539d685119e16ea8782838723e2bc5d1/workspace/build/elf/ld.so.new      \
      -Wl,-z,combreloc -Wl,-z,relro -Wl,--hash-style=both -Wl,-z,defs   \
      /bob/86066a2b539d685119e16ea8782838723e2bc5d1/workspace/build/elf/librtld.os -Wl,--version-script=/bob/86066a2b539d685119e16ea8782838723e2bc5d1/workspace/build/ld.map    \
      -Wl,-soname=ld-linux-aarch64.so.1    \
      -Wl,-defsym=_begin=0
/toolchain/lib/gcc/aarch64-bob-linux-gnu/9.2.0/../../../../aarch64-bob-linux-gnu/bin/ld: /bob/86066a2b539d685119e16ea8782838723e2bc5d1/workspace/build/elf/librtld.os: relocation R_AARCH64_ABS32 against `a local symbol' can not be used when making a shared object
/toolchain/lib/gcc/aarch64-bob-linux-gnu/9.2.0/../../../../aarch64-bob-linux-gnu/bin/ld: /bob/86066a2b539d685119e16ea8782838723e2bc5d1/workspace/build/elf/librtld.os: relocation R_AARCH64_ADR_PREL_PG_HI21 against symbol `_dl_argv' which may bind externally can not be used when making a shared object; recompile with -fPIC
/bob/86066a2b539d685119e16ea8782838723e2bc5d1/workspace/build/elf/librtld.os: in function `_dl_start_user':
(.text+0x64): dangerous relocation: unsupported relocation
/toolchain/lib/gcc/aarch64-bob-linux-gnu/9.2.0/../../../../aarch64-bob-linux-gnu/bin/ld: BFD (GNU Binutils) 2.35.1 assertion fail elfnn-aarch64.c:4853
/toolchain/lib/gcc/aarch64-bob-linux-gnu/9.2.0/../../../../aarch64-bob-linux-gnu/bin/ld: BFD (GNU Binutils) 2.35.1 assertion fail elfnn-aarch64.c:4853
/bob/86066a2b539d685119e16ea8782838723e2bc5d1/workspace/build/elf/librtld.os: in function `do_lookup_unique':
/bob/62d2054d522222a7f0518640dddd7e374c288ef8/workspace/elf/dl-lookup.c:318:(.text+0x825c): relocation truncated to fit: R_AARCH64_LD64_GOT_LO12_NC against symbol `free' defined in .text section in /bob/86066a2b539d685119e16ea8782838723e2bc5d1/workspace/build/elf/librtld.os
/toolchain/lib/gcc/aarch64-bob-linux-gnu/9.2.0/../../../../aarch64-bob-linux-gnu/bin/ld: BFD (GNU Binutils) 2.35.1 assertion fail elfnn-aarch64.c:4853
/toolchain/lib/gcc/aarch64-bob-linux-gnu/9.2.0/../../../../aarch64-bob-linux-gnu/bin/ld: BFD (GNU Binutils) 2.35.1 assertion fail elfnn-aarch64.c:4853
/bob/62d2054d522222a7f0518640dddd7e374c288ef8/workspace/elf/dl-lookup.c:287:(.text+0x8464): relocation truncated to fit: R_AARCH64_LD64_GOT_LO12_NC against symbol `free' defined in .text section in /bob/86066a2b539d685119e16ea8782838723e2bc5d1/workspace/build/elf/librtld.os
/toolchain/lib/gcc/aarch64-bob-linux-gnu/9.2.0/../../../../aarch64-bob-linux-gnu/bin/ld: BFD (GNU Binutils) 2.35.1 assertion fail elfnn-aarch64.c:4853
/toolchain/lib/gcc/aarch64-bob-linux-gnu/9.2.0/../../../../aarch64-bob-linux-gnu/bin/ld: BFD (GNU Binutils) 2.35.1 assertion fail elfnn-aarch64.c:4853
/bob/86066a2b539d685119e16ea8782838723e2bc5d1/workspace/build/elf/librtld.os: in function `elf_machine_runtime_setup':
/bob/62d2054d522222a7f0518640dddd7e374c288ef8/workspace/elf/../sysdeps/aarch64/dl-machine.h:101:(.text+0x9ce0): relocation truncated to fit: R_AARCH64_LD64_GOT_LO12_NC against symbol `_dl_runtime_profile' defined in .text section in /bob/86066a2b539d685119e16ea8782838723e2bc5d1/workspace/build/elf/librtld.os
/toolchain/lib/gcc/aarch64-bob-linux-gnu/9.2.0/../../../../aarch64-bob-linux-gnu/bin/ld: BFD (GNU Binutils) 2.35.1 assertion fail elfnn-aarch64.c:4853
/toolchain/lib/gcc/aarch64-bob-linux-gnu/9.2.0/../../../../aarch64-bob-linux-gnu/bin/ld: BFD (GNU Binutils) 2.35.1 assertion fail elfnn-aarch64.c:4853
/bob/62d2054d522222a7f0518640dddd7e374c288ef8/workspace/elf/../sysdeps/aarch64/dl-machine.h:114:(.text+0xa4ac): relocation truncated to fit: R_AARCH64_LD64_GOT_LO12_NC against symbol `_dl_runtime_resolve' defined in .text section in /bob/86066a2b539d685119e16ea8782838723e2bc5d1/workspace/build/elf/librtld.os
/toolchain/lib/gcc/aarch64-bob-linux-gnu/9.2.0/../../../../aarch64-bob-linux-gnu/bin/ld: BFD (GNU Binutils) 2.35.1 assertion fail elfnn-aarch64.c:4853
/toolchain/lib/gcc/aarch64-bob-linux-gnu/9.2.0/../../../../aarch64-bob-linux-gnu/bin/ld: BFD (GNU Binutils) 2.35.1 assertion fail elfnn-aarch64.c:4853
/bob/86066a2b539d685119e16ea8782838723e2bc5d1/workspace/build/elf/librtld.os: in function `_dl_debug_initialize':
/bob/62d2054d522222a7f0518640dddd7e374c288ef8/workspace/elf/dl-debug.c:51:(.text+0xd694): relocation truncated to fit: R_AARCH64_LD64_GOT_LO12_NC against symbol `_r_debug' defined in COMMON section in /bob/86066a2b539d685119e16ea8782838723e2bc5d1/workspace/build/elf/librtld.os
/toolchain/lib/gcc/aarch64-bob-linux-gnu/9.2.0/../../../../aarch64-bob-linux-gnu/bin/ld: BFD (GNU Binutils) 2.35.1 assertion fail elfnn-aarch64.c:4853
/toolchain/lib/gcc/aarch64-bob-linux-gnu/9.2.0/../../../../aarch64-bob-linux-gnu/bin/ld: BFD (GNU Binutils) 2.35.1 assertion fail elfnn-aarch64.c:4853
/bob/62d2054d522222a7f0518640dddd7e374c288ef8/workspace/elf/dl-debug.c:59:(.text+0xd6b0): relocation truncated to fit: R_AARCH64_LD64_GOT_LO12_NC against symbol `_r_debug' defined in COMMON section in /bob/86066a2b539d685119e16ea8782838723e2bc5d1/workspace/build/elf/librtld.os
/toolchain/lib/gcc/aarch64-bob-linux-gnu/9.2.0/../../../../aarch64-bob-linux-gnu/bin/ld: BFD (GNU Binutils) 2.35.1 assertion fail elfnn-aarch64.c:4853
/toolchain/lib/gcc/aarch64-bob-linux-gnu/9.2.0/../../../../aarch64-bob-linux-gnu/bin/ld: BFD (GNU Binutils) 2.35.1 assertion fail elfnn-aarch64.c:4853
/bob/86066a2b539d685119e16ea8782838723e2bc5d1/workspace/build/elf/librtld.os: in function `htab_expand':
/bob/62d2054d522222a7f0518640dddd7e374c288ef8/workspace/elf/../include/inline-hashtab.h:175:(.text+0x13308): relocation truncated to fit: R_AARCH64_LD64_GOT_LO12_NC against symbol `free' defined in .text section in /bob/86066a2b539d685119e16ea8782838723e2bc5d1/workspace/build/elf/librtld.os
/toolchain/lib/gcc/aarch64-bob-linux-gnu/9.2.0/../../../../aarch64-bob-linux-gnu/bin/ld: BFD (GNU Binutils) 2.35.1 assertion fail elfnn-aarch64.c:4853
/toolchain/lib/gcc/aarch64-bob-linux-gnu/9.2.0/../../../../aarch64-bob-linux-gnu/bin/ld: BFD (GNU Binutils) 2.35.1 assertion fail elfnn-aarch64.c:4853
/bob/86066a2b539d685119e16ea8782838723e2bc5d1/workspace/build/elf/librtld.os: in function `htab_create':
/bob/62d2054d522222a7f0518640dddd7e374c288ef8/workspace/elf/../include/inline-hashtab.h:56:(.text+0x13344): relocation truncated to fit: R_AARCH64_LD64_GOT_LO12_NC against symbol `free' defined in .text section in /bob/86066a2b539d685119e16ea8782838723e2bc5d1/workspace/build/elf/librtld.os
/bob/86066a2b539d685119e16ea8782838723e2bc5d1/workspace/build/elf/librtld.os:(.data.rel.ro+0x178): dangerous relocation: unsupported relocation
/bob/86066a2b539d685119e16ea8782838723e2bc5d1/workspace/build/elf/librtld.os:(.data.rel.ro+0x180): dangerous relocation: unsupported relocation
/bob/86066a2b539d685119e16ea8782838723e2bc5d1/workspace/build/elf/librtld.os:(.data.rel.ro+0x188): dangerous relocation: unsupported relocation
/bob/86066a2b539d685119e16ea8782838723e2bc5d1/workspace/build/elf/librtld.os:(.data.rel.ro+0x190): dangerous relocation: unsupported relocation
/bob/86066a2b539d685119e16ea8782838723e2bc5d1/workspace/build/elf/librtld.os:(.data.rel.ro+0x198): dangerous relocation: unsupported relocation
/bob/86066a2b539d685119e16ea8782838723e2bc5d1/workspace/build/elf/librtld.os:(.data.rel.ro+0x1a0): dangerous relocation: unsupported relocation
/bob/86066a2b539d685119e16ea8782838723e2bc5d1/workspace/build/elf/librtld.os:(.data.rel.ro+0x1a8): dangerous relocation: unsupported relocation
/bob/86066a2b539d685119e16ea8782838723e2bc5d1/workspace/build/elf/librtld.os:(.data.rel.ro+0x1b0): dangerous relocation: unsupported relocation
/bob/86066a2b539d685119e16ea8782838723e2bc5d1/workspace/build/elf/librtld.os:(.data.rel.ro+0x1b8): dangerous relocation: unsupported relocation
/bob/86066a2b539d685119e16ea8782838723e2bc5d1/workspace/build/elf/librtld.os:(.data.rel.ro+0x1c0): dangerous relocation: unsupported relocation
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:380: /bob/86066a2b539d685119e16ea8782838723e2bc5d1/workspace/build/elf/ld.so] Error 1
make[2]: Leaving directory '/bob/62d2054d522222a7f0518640dddd7e374c288ef8/workspace/elf'
make[1]: *** [Makefile:214: elf/subdir_lib] Error 2
make[1]: Leaving directory '/bob/62d2054d522222a7f0518640dddd7e374c288ef8/workspace'
make: *** [Makefile:9: all] Error 2
++ bob_handle_error 2
++ set +x
Step failed with return status 2; Command: make -j "${MAKE_JOBS-$(nproc)}" "$@"
Call stack (most recent call first)
    #0: Class  make, layer self, line 9, in makeParallel
    #1: Recipe libs::compat::glibc#1, layer self, line 20, in main
Build error: Build script ./dev/build/libs/compat/glibc-dev/1/build.sh returned with 2
Failed package: cmake::greeter-cross/devel::host-compat-toolchain/devel::compat::cross-toolchain/devel::compat::gcc-cross/libs::compat::glibc-dev
```

**Comment by mahaase** [comment]
i guess this is a bad idea, that i try this xD
the modules seams not ready for aarch64.

first i had some easy problems in modules like expat or libcap. than glibc, i searched hours for the patches.
now the gcc-cross breaks :-/

but the aarch64-sandbox itself works :)
did u think it is possible to get this up?

File: recipes/devel/host-compat-toolchain.yaml [hunk 3]
```diff
@@ -6,7 +6,7 @@
 shared: True
 
 environment:
-    AUTOCONF_TARGET: "$(gen-autoconf,bob_compat)"
+    AUTOCONF_TARGET: "${SANDBOX_AUTOCONF_TARGET:-$(gen-autoconf,bob_compat)}"
```
Comment (mahaase) [inline recipes/devel/host-compat-toolchain.yaml:9]: this is wrong! host-compat doesn't match, if we have a cross-compiled sandbox!
that was not a requirement so far. maybe this would be work:

`${AUTOCONF_TARGET:-$(gen-autoconf,bob_compat)}`

use host, if not already set...?!

**Comment by jkloetzke** [comment]
> As always, I'm more than confused when it comes to this `host-toolchain` vs. `target-toolchain` stuff. From my perspective it looks good, but I'd like to wait for @jkloetzke's opinion.

Actually it is quite simple:  `host-toolchain` builds for the system where Bob is running on and `target-toolchain` for, well, the target system. If you are not cross compiling both are identical. In the cross-compiling use case they differ.

> but the aarch64-sandbox itself works :)
> did u think it is possible to get this up?

I could imagine that it's possible but it would be good to define precisely what you want to accomplish. If I understand correctly then you want to:

* create a sandbox that has a foreign architecture
* put a statically linked qemu inside the sandbox to translate from the foreign architecture to the host
* have some kind of native compiler in the sandbox

Is this correct?

> makes it sens to build a `BASEMENT_HOST_COMPAT_TOOLCHAIN` if we are inside the sandbox?

Depends on you use case where you want to execute Bob. Does the sandbox run on the target? Will it be the same glibc? Or does the toolchain need to run on other Distros too?

**Comment by mahaase** [comment]
> put a statically linked qemu inside the sandbox to translate from the foreign architecture to the host

how ever this is not necessary. for me, it works out of the box (with the tool installed on my machine).
if u use this branch and build the sandbox successfully, u can still use it, e.g. by calling the `build.sh shell`, et voila u are inside a aarch64 sandbox system.

my use-cases i want to bring into the bob:
(+) x86_64 native builds (windows + linux)
(+) x86_64 cross builds (to newer gcc + glibc) with e.g. a docker image in the end to execute the stuff (windows + linux)
(+) x86_64 to aarch64 cross build (our targets are nvidia agx embedded systems) (linux)
(-) aarch64 native build: (linux)
      aarch64 sandbox (e.g. on x86_64 jenkins to create artifafcts, which can be used by native nvidia targets)
      on the aarch64 nvidia target itself

* all with and without cuda (windows + linux)

> Does the sandbox run on the target? Will it be the same glibc? Or does the toolchain need to run on other Distros too?

that's the point! we need the `host-compat-toolchain` because embedded target != sandbox. but we call bob on host (with aarch64 sandbox) and on target without sandbox!

but the glibc problem is a general one! ALL our linux machines (embedded and pc) are ubuntu 18 lts. the current basement target toolchain is still to new for us. 

bionic: glibc 2.27; gcc 3.3 to 8; cuda 10.2 (which also supports just gcc until v8)

**Comment by mahaase** [comment]
if i get this:

```
rpc_clntout.c:32:19: fatal error: stdio.h: No such file or directory
compilation terminated.
rpc_hout.c:36:19: fatal error: stdio.h: No such file or directory
compilation terminated.
rpc_sample.c:37:19: fatal error: stdio.h: No such file or directory
compilation terminated.
rpc_svcout.c:36:19: fatal error: stdio.h: No such file or directory
rpc_util.c:36:19: fatal error: stdio.h: No such file or directory
```

while building: `cmake::greeter-cross-aarch64/devel::host-compat-toolchain/devel::compat::cross-toolchain/devel::compat::gcc-cross/libs::compat::glibc-dev`

than there is an error in the sandbox environment, isn't it? we missed the installation of the headers to /usr/include...
for a x86_64 sandbox, does the system use the host headers?

**Comment by jkloetzke** [comment]
I'm looking into it but it will take some time.

> bionic: glibc 2.27; gcc 3.3 to 8; cuda 10.2 (which also supports just gcc until v8)

I hope gcc 3.3 is a typo. Ubuntu bionic should come with gcc 7.

> if the sandbox will be cross-compiled like here, what is the correct way? Also cross-compiling the bootstrap-sandbox, or just the sandbox?

I think the sandbox cross compiling is correct. The bootstrap-sandbox is just an extra layer to make the real sandbox build reproducible. Therefore it should stay in the native architecture.

**Comment by mahaase** [comment]
> I hope gcc 3.3 is a typo. Ubuntu bionic should come with gcc 7.

there is a range of installable gcc verions pro ubuntu release. for 18lts there are gcc 4.8 to 8, but indeed not 3.3! default is 7!

I pushed my last ideas, that the sandbox provides its information into `AUTOCONF_HOST` and `AUTOCONF_BUILD`.


> I'm looking into it but it will take some time.

no pressure! every help is welcome! :)

File: recipes/libs/compat/glibc/0003-improve-ifunc-check-BZ-25506.diff.save [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
++++ b/configure.ac
+@@ -649,7 +649,7 @@ if ${CC-cc} $CFLAGS $CPPFLAGS $LDFLAGS \
+ 	    -o conftest conftest.S 1>&AS_MESSAGE_LOG_FD 2>&AS_MESSAGE_LOG_FD; then
+   # Do a link to see if the backend supports IFUNC relocs.
+   $READELF -r conftest 1>&AS_MESSAGE_LOG_FD
+-  LC_ALL=C $READELF -r conftest | grep 'no relocations' >/dev/null || {
++  LC_ALL=C $READELF -Wr conftest | grep -q 'IRELATIVE\|R_SPARC_JMP_IREL' && {
+     libc_cv_ld_gnu_indirect_function=yes
+   }
+ fi
```
Comment (Ferruck) [inline recipes/libs/compat/glibc/0003-improve-ifunc-check-BZ-25506.diff.save:27]: Have you taken this from https://sourceware.org/pipermail/glibc-cvs/2020q1/000002.html ?

If yes, please add the patch with the original author, commit message and a link to the original source (at least in *your* commit message) included, otherwise this might infringe copyrights...

Comment (mahaase) [reply recipes/libs/compat/glibc/0003-improve-ifunc-check-BZ-25506.diff.save]: thanks for that information, but i guess this patch was not needed, i removed it.

**Comment by jkloetzke** [comment]
I think the last changes (--with-linker-hash-style=gnu) should have gone to #84 instead, right?

**Comment by mahaase** [comment]
shit i pushed the merged changes in here... i have to fix this. thanks!

**Comment by mahaase** [comment]
https://github.com/BobBuildTool/basement/blob/fe037ed5d29b92402010b7078c7efb4956730cf4/recipes/devel/compat/gcc.yaml#L137

i do not understand. in my opinion this line is not correct. this path will not exists. to which location shall this reference?
to the location of the gcc-cross or to the build-dir of cross-toolchain or something else?

**Comment by jkloetzke** [comment]
> https://github.com/BobBuildTool/basement/blob/fe037ed5d29b92402010b7078c7efb4956730cf4/recipes/devel/compat/gcc.yaml#L137
> 
> i do not understand. in my opinion this line is not correct. this path will not exists. to which location shall this reference?
> to the location of the `gcc-cross` or to the build-dir of `cross-toolchain` or something else?

Even though it looks like an absolute path it will still be implicitly relocated by gcc. I don't have the documentation at hand but if the configured sysroot is *below* the configured install prefix (which is `/usr`) it will be considered for relocation too. In the end, gcc will look for the sysroot relative to its executable.

The sysroot is actually created in the cross-toolchain recipe when the gcc- and libc-target-files are copied into the sysroot at this location.

**Comment by mahaase** [comment]
i guess, with the last changes the generic sandbox does work. (aarch64 sandbox)
still some retests necessary.

imo the switch from `aarch64-linux-gnu` to `aarch64-cross-linux-gnu` was an important step to bring this up without missing headerfile errors.

**Comment by daxcore** [comment]
uhh. still looks fine for me. i also tested the `cmake::greeter-cross-aarch64` recipe. a aarch64 sandbox with x86_64 target compiler. still works. :)

**Comment by mahaase** [comment]
the most problems already fixed. there are just some adaptions left. hopefully we can bring this feature to master soon 😃

**Comment by mahaase** [comment]
some important information i will add to the commit message finally:

the "generic sandbox" feature is also working with WSL2 (up to ubuntu 18 lts tested)
* `sudo apt install qemu binfmt-support qemu-user-static`
* install a newer version of qemu-user-static (>=5) manually
  - https://packages.ubuntu.com/kinetic/amd64/qemu-user-static/download
  - `sudo dpkg -i <filename>.deb`
* workaround for WSL2 Ubuntu 18 LTS only (not needed on WSL2 Ubuntu 20 LTS or native Ubuntu >= 18 LTS): 
  - `sudo update-binfmts --unimport qemu-aarch64`
  - `sudo update-binfmts --import qemu-aarch64`
* `sudo /etc/init.d/binfmt-support (re-)start` or if available: `sudo systemctl (re-)start binfmt-support.service`
* important, check that `update-binfmts --display` shows `interpreter = /usr/libexec/qemu-binfmt/*` instead of` /usr/bin/qemu-*-static` directly

**Comment by mahaase** [comment]
close. follow up here: #154
